pkg/alertmanager/statefulset.go

```diff
@@ -210,7 +210,7 @@ func makeStatefulSetSpec(a *monitoringv1.Alertmanager, config Config) (*appsv1.S
 	// details see https://github.com/prometheus-operator/prometheus-operator/issues/1659
 	a = a.DeepCopy()
 
-	amBaseImage := operator.StringValOrDefault(a.Spec.BaseImage, operator.DefaultAlertmanagerBaseImage)
+	amBaseImage := operator.StringValOrDefault(a.Spec.BaseImage, config.AlertmanagerDefaultBaseImage)
```

Comment:
I've noticed that we're not consistent across the code base on when we call `operator.BuildImagePath()`: like here we call it before overriding the default image's path with `Spec.Image` (if it isn't empty), in other places like [prometheus](https://github.com/prometheus-operator/prometheus-operator/blob/30b03e839a460ef230256873c10ad56f9a5abd37/pkg/prometheus/statefulset.go#L320-L327) we call `operator.BuildImagePath()` after overriding.

cc @brancz @s-urbaniak @lilic

Reply:
Interesting, yeah the `operator.BuildImagePath ` is meant to be used to create the image path so we should be consistent in using it, agreed.

Reply:
Prometheus:
https://github.com/prometheus-operator/prometheus-operator/blob/30b03e839a460ef230256873c10ad56f9a5abd37/pkg/prometheus/statefulset.go#L320-L324

Additional Thanos in Prometheues:
https://github.com/prometheus-operator/prometheus-operator/blob/30b03e839a460ef230256873c10ad56f9a5abd37/pkg/prometheus/statefulset.go#L686-L698

Alertmanager:
https://github.com/prometheus-operator/prometheus-operator/blob/30b03e839a460ef230256873c10ad56f9a5abd37/pkg/alertmanager/statefulset.go#L213-L223

Thanos:
https://github.com/prometheus-operator/prometheus-operator/blob/30b03e839a460ef230256873c10ad56f9a5abd37/pkg/thanos/statefulset.go#L155-L162

Which pattern above is to be used across all?
SHA and Tag is not available for the Thanos - not sure why?

Reply:
Thanos doesn't have SHA and tag values because these fields have been deprecated for Alertmanager and Prometheus before the Thanos Ruler CRD exists.

This is the API documentation of the `Image` field:

```
Image if specified has precedence over baseImage, tag and sha combinations. Specifying the version is still necessary to ensure the Prometheus Operator knows what version of Prometheus is being configured.
```

I'd say that the operator should pick the value of `Image` unmodified if it isn't empty otherwise it should construct the image path from `BaseImage` + `SHA` + `Tag`. Which would be more or less the implementation for Alertmanager except that it shouldn't even call `BuildImagePath()` if `Image` isn't empty. But lets hear what other maintainers think :)

Reply:
gentle ping to @lilic @brancz @s-urbaniak Please let me know your agreement or otherwise with @simonpasquier comments above. I can make the changes and commit them.

Reply:
> Which would be more or less the implementation for Alertmanager except that it shouldn't even call BuildImagePath() if Image isn't empty.

yes that sounds good for me, we could just put that logic in the BuildImagePath function and just call that, or is there anything specific to Alertmanager @simonpasquier ?

Reply:
> is there anything specific to Alertmanager?

Not that I know of.

Reply:
Then I would prefer if we do it the same for both Prometheus and alertmanager, so lets adjust the BuildImagePath function, does that sound good?

Reply:
I have made a new function in image.go. Please check if its fine. I can make changes in other places as well and submit the fix for final apporval

Reply:
I think that @lilic's idea was to pass `a.Spec.Image` as an additional parameter to `BuildImagePath()` and return it unmodified if it isn't empty. The rest of `BuildImagePath()` would stay the same. And in this code we can remove any `if` statements:

```
amImagePath, err := operator.BuildImagePath(
    operator.StringPtrValOrDefault(a.Spec.Image, ""), // image path from path.
    operator.StringValOrDefault(a.Spec.BaseImage, config.AlertmanagerDefaultBaseImage), // base image path from spec or from operator's config.
    operator.StringValOrDefault(a.Spec.Version, operator.DefaultAlertmanagerVersion), // version from spec or operator's default.
    operator.StringValOrDefault(a.Spec.Tag, ""), // tag from spec.
    operator.StringValOrDefault(a.Spec.SHA, ""), // SHA from spec.
)
```

Reply:
Passing a.Spec.Image as additional argument is fine with me. I think that by passing additional arguments, we can avoid calling operator.StringValOrDefault till the time its really made use of. Does that make sense?

Reply:
Yes sorry for not being specific enough, exactly what Simon said is what I had in mind. This would be the central function to decide on image building path then.

Reply:
The types for members which have same name in different specs are NOT consistent between different specs. This is making the `BuildImagePath `to be not useful consistently and also making code a little messy.
Example
Image is **pointer to string** in PrometheusSpec , but **plain string** in ThanosRulerSpec .
Similarly, BaseImage is **plain string** in PrometheusSpec , but **pointer to string** in ThanosSpec.
ThanosSpec has everything as pointer to string :confused:
```
type PrometheusSpec struct {
Version string
Tag string
SHA string
Image *string
BaseImage string
....
}
```

```
type AlertmanagerSpec struct {
Image *string
Version string
Tag string
SHA string
BaseImage string
...
}
```
```
type ThanosRulerSpec struct {
	Image string
...
```

```
type ThanosSpec struct {
Image *string
Version *string
Tag *string
SHA *string
BaseImage *string
...
}
```
I would propose to modify all Spec to have one set of types aligned to PrometheusSpec.
AlertManagerSpec is already aligned.

```
type ThanosRulerSpec struct {
	Image *string
...
```

```
type ThanosSpec struct {
Image *string
Version string
Tag string
SHA string
BaseImage string
...
}
```
With this, BuildImagePath will be consistently usable. Will this have any side effects? I can make the changes and see whether CI tests pass or not. If there are any early WARNINGs please let me know. :hand:

Reply:
`BuildImagePath()` only deals with values and not pointers. I'm fine with the caller doing the conversion using `operator.StringPtrValOrDefault() or `operator.StringValOrDefault()` as needed. Does it make sense?

Reply:
Fine.  Let me make changes accordingly. I would still feel that aligning all the types across spec's will make code even simpler and consistent. It can be done in another PR. I can create an issue for the same if there is agreement.

Reply:
If by spec you mean the user facing API, we cannot and should not change those, we need to workaround those in code.

Reply:
By spec, I am referring to types defined in pkg/apis/monitoring/v1/types.go for PrometheusSpec, AlertManagerSpec, and ThanosSpec